src/rust/cryptography-x509/src/common.rs

```diff
@@ -301,9 +301,72 @@ impl<'a> asn1::SimpleAsn1Writable for UnvalidatedVisibleString<'a> {
     }
 }
 
+/// Like `UnvalidatedIA5String`, but preserves the invariant that the
+/// underlying string is ASCII only.
+#[derive(Debug, PartialEq)]
+pub struct IA5String(String);
+
+impl IA5String {
+    pub(crate) fn new(value: String) -> Option<Self> {
+        if value.is_ascii() {
+            Some(Self(value))
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+/// A `DNSString` is an `IA5String` with additional invariant preservations
+/// per RFC 5280 4.2.1.6.
+///
+/// In particular, a `DNSString` is normalized to lowercase ASCII internally
+/// and cannot contain `" "`.
+#[derive(Debug, PartialEq)]
+pub struct DNSString(IA5String);
+
+impl DNSString {
+    pub fn new(value: &str) -> Option<Self> {
+        if value.is_empty() || value == " " {
+            None
+        } else {
+            IA5String::new(value.to_lowercase()).map(Self)
+        }
+    }
```

Comment:
Question for @reaperhulk and @alex: how strong would you like invariant preservation to be here? Right now we're just checking for ASCII-ness and a few things that 5280 explicitly disallows, but we could make this into a full DNS label/domain validation check instead (which would be similar to what OpenSSL does.)